File: medium/segment trees/range-sum-query-mutable.py
# name: 307. range sum query - mutable
# link: https://leetcode.com/problems/range-sum-query-mutable

# solution #
class SegmentNode:
	def __init__(self, total=0, L=0, R=0):
		self.sum = total
		self.L, self.R = L, R
		self.left = self.right = None
	
	def build(self, nums, L, R):
		if L == R:
			return SegmentNode(nums[L], L, R)
		
		M = (L + R) // 2
		
		# form the tree
		root = SegmentNode(0, L, R)
		root.left = self.build(nums, L, M)
		root.right = self.build(nums, M + 1, R)
		
		# store the range sum
		root.sum = root.left.sum + root.right.sum
		return root
	
	def update(self, idx, val):
		# at a given index
		if self.L == self.R:
			self.sum = val
			return
		
		M = (self.L + self.R) // 2
		
		if idx <= M:
			self.left.update(idx, val)
		else:
			self.right.update(idx, val)
		
		# update the range sum
		self.sum = self.left.sum + self.right.sum
	
	def rangeQuery(self, L, R):
		if self.L == L and self.R == R:
			return self.sum
		
		M = (self.L + self.R) // 2
		
		if L > M:
			return self.right.rangeQuery(L, R)
		elif R <= M:
			return self.left.rangeQuery(L, R)
		else:
			return self.left.rangeQuery(L, M) + self.right.rangeQuery(M + 1, R)

class NumArray:
	
	def __init__(self, nums: List[int]):
		self.root = SegmentNode()
		# build the segment tree
		self.root = self.root.build(nums, 0, len(nums) - 1)
	
	def update(self, index: int, val: int) -> None:
		self.root.update(index, val)
	
	def sumRange(self, left: int, right: int) -> int:
		return self.root.rangeQuery(left, right)

"""
time complexity:
- O(n)

space complexity:
- O(n)
"""

File: medium/segment trees/range-sum-query-mutable.go
package segment_trees

// name: 307. range sum query - mutable
// link: https://leetcode.com/problems/range-sum-query-mutable

// SegmentNode solution //
type SegmentNode struct {
	Sum         int
	L, R        int
	Left, Right *SegmentNode
}

func (sn *SegmentNode) Build(nums []int, L, R int) *SegmentNode {
	if L == R {
		return &SegmentNode{Sum: nums[L], L: L, R: R}
	}

	M := (L + R) / 2

	// build the tree
	root := &SegmentNode{L: L, R: R}
	root.Left = sn.Build(nums, L, M)
	root.Right = sn.Build(nums, M+1, R)

	root.Sum = root.Left.Sum + root.Right.Sum
	return root
}

func (sn *SegmentNode) Update(index, val int) {
	if sn.L == sn.R {
		sn.Sum = val
		return
	}

	M := (sn.L + sn.R) / 2

	if index <= M {
		sn.Left.Update(index, val)
	} else {
		sn.Right.Update(index, val)
	}

	sn.Sum = sn.Left.Sum + sn.Right.Sum
}

func (sn *SegmentNode) RangeQuery(L, R int) int {
	if sn.L == L && sn.R == R {
		return sn.Sum
	}

	M := (sn.L + sn.R) / 2

	if L > M {
		return sn.Right.RangeQuery(L, R)
	} else if R <= M {
		return sn.Left.RangeQuery(L, R)
	} else {
		return sn.Left.RangeQuery(L, M) + sn.Right.RangeQuery(M+1, R)
	}
}

type NumArray struct {
	root *SegmentNode
}

func Constructor(nums []int) NumArray {
	dummy := SegmentNode{}
	root := dummy.Build(nums, 0, len(nums)-1)
	return NumArray{root: root}
}

func (na *NumArray) Update(index int, val int) {
	na.root.Update(index, val)
}

func (na *NumArray) SumRange(left int, right int) int {
	return na.root.RangeQuery(left, right)
}

/**
time complexity:
- O(n)

space complexity:
- O(n)
*/
